bytecomp/lambda.mli

```diff
@@ -332,9 +332,15 @@ val name_lambda: let_kind -> lambda -> (Ident.t -> lambda) -> lambda
 val name_lambda_list: lambda list -> (lambda list -> lambda) -> lambda
 
 val iter_head_constructor: (lambda -> unit) -> lambda -> unit
-(** [iter_head_constructor f lam] apply [f] to only the first level of
+(** [iter_head_constructor f lam] apply [f] or [tail] to only the first level of
     sub expressions of [lam]. It does not recursively traverse the
-    expression. *)
+    expression.
+*)
+
+val shallow_iter: tail:(lambda -> unit) -> non_tail:(lambda -> unit) -> lambda -> unit
```

Comment:
This function is fine, but personally I think that `(tail:bool -> lambda -> unit) -> lambda -> unit` would be a bit easier to use than the current proposal. It makes it easy to ignore the information (you can completely replace the main `iter` function with this generalization, instead of having to provide a separate function because the API is a bit inconvenient), and I think in most cases `tail` and `non_tail` are going to share most of their logic, so marking the differences with `if tail ...;` inserts makes a lot of sense.

Reply:
> you can completely replace the main iter function with this generalization, instead of having to provide a separate function because the API is a bit inconvenient

The current approach allows defining the previous `iter_head_constructor` (which does not distinguish between the two kinds of sub-nodes) without having to allocate a closure at each node; and for the only use of the more fine-grained version (in Simplif), it works quite well to have two different callbacks.